src/main/java/dev/morling/onebrc/CalculateAverage_vaidhy.java

```diff
@@ -0,0 +1,427 @@
+                lineStream.skipLine();
+            }
+            else {
+                // No lines then do nothing.
+                return;
+            }
+        }
+        while (lineStream.hasNext()) {
+            long keyStartAddress = lineStream.position;
+            long keyEndAddress = lineStream.findSemi();
+            long keySuffix = lineStream.suffix;
+            int keyHash = lineStream.hash;
+            long valueStartAddress = lineStream.position;
+            long valueEndAddress = lineStream.findTemperature();
+            int temperature = parseDouble(valueStartAddress, valueEndAddress);
+            lineConsumer.process(keyStartAddress, keyEndAddress, keyHash, temperature, keySuffix);
+        }
+    }
+
+    public T master(long chunkSize, ExecutorService executor) {
+        long len = fileService.length();
+        List<Future<I>> summaries = new ArrayList<>();
+
+        for (long offset = 0; offset < len; offset += chunkSize) {
+            long workerLength = Math.min(len, offset + chunkSize) - offset;
+            MapReduce<I> mr = chunkProcessCreator.get();
+            final long transferOffset = offset;
+            Future<I> task = executor.submit(() -> {
+                worker(transferOffset, workerLength, mr);
+                return mr.result();
+            });
+            summaries.add(task);
+        }
+        List<I> summariesDone = summaries.stream()
+                .map(task -> {
+                    try {
+                        return task.get();
+                    }
+                    catch (InterruptedException | ExecutionException e) {
+                        throw new RuntimeException(e);
+                    }
+                })
+                .toList();
+        return reducer.apply(summariesDone);
+    }
+
+    static class DiskFileService implements FileService {
+        private final long fileSize;
+        private final long mappedAddress;
+
+        DiskFileService(String fileName) throws IOException {
+            FileChannel fileChannel = FileChannel.open(Path.of(fileName),
+                    StandardOpenOption.READ);
+            this.fileSize = fileChannel.size();
+            this.mappedAddress = fileChannel.map(FileChannel.MapMode.READ_ONLY, 0,
+                    fileSize, Arena.global()).address();
+        }
+
+        @Override
+        public long length() {
+            return fileSize;
+        }
+
+        @Override
+        public long address() {
+            return mappedAddress;
+        }
+    }
+
+    private static class ChunkProcessorImpl implements MapReduce<PrimitiveHashMap> {
+
+        // 1 << 14 > 10,000 so it works
+        private final PrimitiveHashMap statistics = new PrimitiveHashMap(14);
+
+        @Override
+        public void process(long keyStartAddress, long keyEndAddress, int hash, int temperature, long suffix) {
+            HashEntry entry = statistics.find(keyStartAddress, keyEndAddress, suffix, hash);
```

Comment:
Where do you deal with hash collisions?

Reply:
In the compareEntry method inside statistics.find. In line 68, we check if there is a hash collision, and if so, it goes on to compare the actual key before returning from find.

Reply:
Ah yes. Thx for clarifying!